File: app/src/main/java/social/entourage/android/map/tour/join/TourJoinRequestFragment.java
package social.entourage.android.map.tour.join;

import android.content.Context;
import android.graphics.Color;
import android.graphics.drawable.ColorDrawable;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.app.DialogFragment;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.Window;
import android.view.WindowManager;
import android.view.inputmethod.InputMethodManager;
import android.widget.EditText;
import android.widget.TextView;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import social.entourage.android.Constants;
import social.entourage.android.EntourageApplication;
import social.entourage.android.EntourageComponent;
import social.entourage.android.EntourageEvents;
import social.entourage.android.R;
import social.entourage.android.api.model.TimestampedObject;
import social.entourage.android.api.model.map.Entourage;
import social.entourage.android.api.model.map.FeedItem;
import social.entourage.android.api.model.map.Tour;

public class TourJoinRequestFragment extends DialogFragment {

    // ----------------------------------
    // CONSTANTS
    // ----------------------------------

    public static final String TAG = "tour_join_request_message";

    // ----------------------------------
    // ATTRIBUTES
    // ----------------------------------

    @BindView(R.id.tour_join_request_ok_description)
    TextView descriptionTextView;

    @BindView(R.id.tour_join_request_ok_message)
    EditText messageView;

    // ----------------------------------
    // PRIVATE MEMBERS
    // ----------------------------------

    private FeedItem feedItem;

    @Inject
    TourJoinRequestPresenter presenter;

    private boolean startedTyping = false;

    // ----------------------------------
    // LIFECYCLE
    // ----------------------------------

    public TourJoinRequestFragment() {
        // Required empty public constructor
    }


    public static TourJoinRequestFragment newInstance(FeedItem feedItem) {
        TourJoinRequestFragment fragment = new TourJoinRequestFragment();
        Bundle args = new Bundle();
        args.putSerializable(Tour.KEY_TOUR, feedItem);
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        //setStyle(DialogFragment.STYLE_NO_TITLE, android.R.style.Theme_Holo_Light_Dialog_NoActionBar_MinWidth);
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {

        if (getArguments() != null) {
            feedItem = (FeedItem) getArguments().getSerializable(Tour.KEY_TOUR);
        }
        if (getDialog() != null && getDialog().getWindow() != null) {
            getDialog().getWindow().requestFeature(Window.FEATURE_NO_TITLE);
            getDialog().getWindow().setBackgroundDrawable(new ColorDrawable(Color.TRANSPARENT));
            getDialog().getWindow().setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_ADJUST_RESIZE | WindowManager.LayoutParams.SOFT_INPUT_STATE_HIDDEN);
        }
        // Inflate the layout for this fragment
        View view = inflater.inflate(R.layout.fragment_tour_join_request_ok, container, false);
        ButterKnife.bind(this, view);

        view.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(final View v) {
                if (getActivity() == null) return;
                InputMethodManager imm = (InputMethodManager)getActivity().getSystemService(Context.INPUT_METHOD_SERVICE);
                if (imm != null) imm.hideSoftInputFromWindow(messageView.getWindowToken(), 0);
            }
        });

        return view;
    }

    @Override
    public void onViewCreated(@NonNull final View view, final Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        setupComponent(EntourageApplication.get().getEntourageComponent());

        if (feedItem != null) {
            int descriptionTextId = R.string.tour_join_request_ok_description_tour;
            if (feedItem.getType() == TimestampedObject.ENTOURAGE_CARD) {
                Entourage entourage = (Entourage)feedItem;
                descriptionTextId = Entourage.TYPE_OUTING.equalsIgnoreCase(entourage.getGroupType()) ? R.string.tour_join_request_ok_description_outing : R.string.tour_join_request_ok_description_entourage;
            }
            descriptionTextView.setText(descriptionTextId);
        }

        messageView.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(final CharSequence s, final int start, final int count, final int after) {

            }

            @Override
            public void onTextChanged(final CharSequence s, final int start, final int before, final int count) {

            }

            @Override
            public void afterTextChanged(final Editable s) {
                if (s.length() > 0 && !startedTyping) {
                    EntourageEvents.logEvent(EntourageEvents.EVENT_JOIN_REQUEST_START);
                    startedTyping = true;
                }
            }
        });
    }

    @OnClick(R.id.tour_join_request_ok_message_button)
    protected void onMessageSend() {
        if (presenter != null && messageView != null) {
            if ( feedItem != null && (feedItem.getType() == FeedItem.TOUR_CARD || feedItem.getType() == FeedItem.ENTOURAGE_CARD) ) {
                EntourageEvents.logEvent(EntourageEvents.EVENT_JOIN_REQUEST_SUBMIT);
                presenter.sendMessage(messageView.getText().toString(), feedItem);
            }
            else {
                dismiss();
            }
        }
    }

    @OnClick(R.id.tour_join_request_ok_x_button)
    protected void onClosePressed() {
        dismiss();
    }

    protected void setupComponent(EntourageComponent entourageComponent) {
        DaggerTourJoinRequestComponent.builder()
                .entourageComponent(entourageComponent)
                .tourJoinRequestModule(new TourJoinRequestModule(this))
                .build()
                .inject(this);
    }
}

File: app/src/main/java/social/entourage/android/entourage/join/TourJoinRequestFragment.kt
package social.entourage.android.entourage.join

import android.content.Context
import android.graphics.Color
import android.graphics.drawable.ColorDrawable
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.*
import android.view.inputmethod.InputMethodManager
import androidx.fragment.app.DialogFragment
import kotlinx.android.synthetic.main.fragment_tour_join_request_ok.*
import social.entourage.android.EntourageApplication.Companion.get
import social.entourage.android.EntourageComponent
import social.entourage.android.EntourageEvents
import social.entourage.android.R
import social.entourage.android.api.model.BaseEntourage
import social.entourage.android.api.model.TimestampedObject
import social.entourage.android.api.model.feed.FeedItem
import social.entourage.android.api.model.tour.Tour
import javax.inject.Inject

class TourJoinRequestFragment  : DialogFragment() {

    // ----------------------------------
    // PRIVATE MEMBERS
    // ----------------------------------
    private lateinit var feedItem: Tour

    @Inject
    lateinit var presenter: TourJoinRequestPresenter

    private var startedTyping = false

    // ----------------------------------
    // LIFECYCLE
    // ----------------------------------
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        dialog?.window?.let {
            it.requestFeature(Window.FEATURE_NO_TITLE)
            it.setBackgroundDrawable(ColorDrawable(Color.TRANSPARENT))
            it.setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_ADJUST_RESIZE or WindowManager.LayoutParams.SOFT_INPUT_STATE_HIDDEN)
        }
        // Inflate the layout for this fragment
        return inflater.inflate(R.layout.fragment_tour_join_request_ok, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        setupComponent(get().entourageComponent)
        val newFeedItem = arguments?.getSerializable(FeedItem.KEY_FEEDITEM) as Tour?
        if(newFeedItem == null) {
            dismiss()
            return
        }
        feedItem = newFeedItem
        var descriptionTextId = R.string.tour_join_request_ok_description_tour
        tour_join_request_ok_description?.setText(descriptionTextId)
        view.setOnClickListener { v: View? ->
            val imm = activity?.getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager? ?:return@setOnClickListener
            tour_join_request_ok_message?.windowToken?.let {
                imm.hideSoftInputFromWindow(it, 0)
            }
        }
        tour_join_request_ok_message?.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {}
            override fun afterTextChanged(s: Editable) {
                if (s.isNotEmpty() && !startedTyping) {
                    EntourageEvents.logEvent(EntourageEvents.EVENT_JOIN_REQUEST_START)
                    startedTyping = true
                }
            }
        })
        tour_join_request_ok_message_button.setOnClickListener {onMessageSend()}
        tour_join_request_ok_x_button.setOnClickListener { dismiss() }
    }

    fun onMessageSend() {
        if (!tour_join_request_ok_message?.text.isNullOrBlank()) {
            EntourageEvents.logEvent(EntourageEvents.EVENT_JOIN_REQUEST_SUBMIT)
            presenter.sendMessage(this, tour_join_request_ok_message!!.text.toString(), feedItem)
        }
    }

    protected fun setupComponent(entourageComponent: EntourageComponent?) {
        DaggerTourJoinRequestComponent.builder()
                .entourageComponent(entourageComponent)
                .build()
                .inject(this)
    }

    companion object {
        // ----------------------------------
        // CONSTANTS
        // ----------------------------------
        const val TAG = "tour_join_request_message"
        fun newInstance(tour: Tour): TourJoinRequestFragment {
            val fragment = TourJoinRequestFragment()
            val args = Bundle()
            args.putSerializable(FeedItem.KEY_FEEDITEM, tour)
            fragment.arguments = args
            return fragment
        }
    }
}